xds/internal/client/load/store.go

```diff
@@ -20,63 +20,121 @@ package load
 import (
 	"sync"
 	"sync/atomic"
+	"time"
 )
 
 const negativeOneUInt64 = ^uint64(0)
 
-// A pair of cluster and service name. The same cluster can be used by multiple
-// services, and one service can use multiple clusters. So we need a pair with
-// both name to accurately indicate where the load belongs.
-type storeKey struct {
-	cluster string
-	service string
-}
-
 // Store keeps the loads for multiple clusters and services to be reported via
-// LRS. It is safe for concurrent use.
+// LRS. It contains loads to reported to one LRS server. Create multiple stores
+// for multiple servers.
+//
+// It is safe for concurrent use.
 type Store struct {
-	mu       sync.RWMutex
-	clusters map[storeKey]*PerClusterStore
+	// mu only protects the map (2 layers). The read/write to *perClusterStore
+	// doesn't need to hold the mu.
+	mu sync.RWMutex
+	// clusters is a map with cluster name as the key. The second layer is a map
+	// with service name as the key. Each value (perClusterStore) contains data
+	// for a (cluster, service) pair.
+	//
+	// Note that new entries are added to this map, but never removed. This is
+	// potentially a memory leak. But the memory is allocated for each new
+	// (cluster,service) pair, and the memory allocated is just pointers and
+	// maps. So this shouldn't get too bad.
+	clusters map[string]map[string]*perClusterStore
 }
 
 // NewStore creates a Store.
 func NewStore() *Store {
 	return &Store{
-		clusters: make(map[storeKey]*PerClusterStore),
+		clusters: make(map[string]map[string]*perClusterStore),
+	}
+}
+
+// Stats returns the load data for the given cluster names. Data is returned in
+// a slice with no specific order.
+//
+// If no clusterName is given (an empty slice), all data for all known clusters
+// are returned.
+//
+// If a cluster's Data is empty (no load to report), it's not append to the
+// returned slice.
+func (s *Store) Stats(clusterNames []string) []*Data {
```

Comment:
I'm wondering if this top-level `Store` should export any methods at all, except for the ones to retrieve the per-cluster store.

Also, I'm wondering whether we should fix the xdsClient API to support reporting loads to multiple clusters in a better way. Things on top of my head (without a lot of serious thought):
- `ReportLoad` which takes `(cluster, service)` should return the `PerClusterStore` and a `cancel` function
- Callers use the returned `PerClusterStore` to report loads, and eventually invoke the `cancel` function
- Internally, the `xdsClient` keeps track of every `cluster+service` combination for which it is reporting load, and the associated `PerClusterStore`
- When it is time to report load (every `loadReporting`  internal), it traverses the list of `PerClusterStore` and fetches stats from them and creates the load report and sends it to the server.